extensions/parquet/base/src/main/java/io/deephaven/parquet/base/ColumnPageReaderImpl.java

```diff
@@ -181,25 +195,14 @@ private DataPageV2 readV2Unsafe(
                 - header.getDefinition_levels_byte_length();
         // With our current single input stream `in` construction, we must copy out the bytes for repetition and
         // definition levels to proceed to data. We could theoretically restructure this to be fully lazy, either by
-        // creating a BytesInput impl for SeekableByteChannel, or by migrating this logic one layer up and ensuring we
+        // creating a ByteBuffer impl for SeekableByteChannel, or by migrating this logic one layer up and ensuring we
         // construct input streams separately for repetitionLevelsIn, definitionLevelsIn, and dataIn. Both of these
         // solutions would potentially suffer from a disconnect in cache-ability that a single input stream provides.
-        final BytesInput repetitionLevels =
-                BytesInput.copy(BytesInput.from(in, header.getRepetition_levels_byte_length()));
-        final BytesInput definitionLevels =
-                BytesInput.copy(BytesInput.from(in, header.getDefinition_levels_byte_length()));
-        final BytesInput data = compressorAdapter.decompress(in, compressedSize, uncompressedSize, channelContext);
-        return new DataPageV2(
-                header.getNum_rows(),
-                header.getNum_nulls(),
-                header.getNum_values(),
-                repetitionLevels,
-                definitionLevels,
-                getEncoding(header.getEncoding()),
-                data,
-                uncompressedPageSize,
-                null, // TODO in the future might want to pull in statistics,
-                false);
+        final ByteBuffer repetitionLevels = readNBytes(in, header.getRepetition_levels_byte_length());
+        final ByteBuffer definitionLevels = readNBytes(in, header.getDefinition_levels_byte_length());
```

Comment:
These could also be read into re-usable cached buffers/byte arrays.
You'd need additional buffers, or the ability to append to the buffer you have. This sort of argues for moving the "key" up to this level, rather than in the decompressor adapter.